kotlin/src/main/kotlin/io/logto/sdk/core/util/GenerateUtils.kt

```diff
@@ -0,0 +1,32 @@
+package io.logto.sdk.core.util
+
+import org.jose4j.base64url.Base64Url
+import java.security.MessageDigest
+
+object GenerateUtils {
+    private const val DEFAULT_ALGORITHM = "SHA-256"
+    private const val DEFAULT_RANDOM_STRING_LENGTH = 64
+
+    fun generateCodeVerifier(): String {
+        return generateRandomString()
+    }
+
+    fun generateCodeChallenge(codeVerifier: String): String {
+        val digester: MessageDigest = MessageDigest.getInstance(DEFAULT_ALGORITHM)
+        digester.update(codeVerifier.toByteArray(Charsets.UTF_8))
+        val byteArray: ByteArray = digester.digest()
+        return Base64Url.encode(byteArray)
+    }
+
+    fun generateState(): String {
+        return generateRandomString()
+    }
+
+    private fun generateRandomString(length: Int = DEFAULT_RANDOM_STRING_LENGTH): String {
+        val actualLen = if (length < 1) DEFAULT_RANDOM_STRING_LENGTH else length
```

Comment:
1. this isn't necessary since the negative cases will be rare?
2. shall we use unsigned int to prevent it from root?
3. naming should be consistent (`actualLength` instead of `actualLen`), based on `length` parameter

Reply:
1. this is a private function, so it's good to remove this line.
2. unsigned int can be 0, so it makes no sense
3. good